homeassistant/components/ads/__init__.py

```diff
@@ -243,16 +268,35 @@ def _device_notification_callback(self, notification, name):
         # Parse data to desired datatype
         if notification_item.plc_datatype == pyads.PLCTYPE_BOOL:
             value = bool(struct.unpack("<?", bytearray(data))[0])
+        elif notification_item.plc_datatype == pyads.PLCTYPE_BYTE:
+            value = struct.unpack("<b", bytearray(data))[0]
         elif notification_item.plc_datatype == pyads.PLCTYPE_INT:
             value = struct.unpack("<h", bytearray(data))[0]
-        elif notification_item.plc_datatype == pyads.PLCTYPE_BYTE:
-            value = struct.unpack("<B", bytearray(data))[0]
         elif notification_item.plc_datatype == pyads.PLCTYPE_UINT:
             value = struct.unpack("<H", bytearray(data))[0]
+        elif notification_item.plc_datatype == pyads.PLCTYPE_SINT:
+            value = struct.unpack("<b", bytearray(data))[0]
+        elif notification_item.plc_datatype == pyads.PLCTYPE_USINT:
+            value = struct.unpack("<B", bytearray(data))[0]
         elif notification_item.plc_datatype == pyads.PLCTYPE_DINT:
             value = struct.unpack("<i", bytearray(data))[0]
         elif notification_item.plc_datatype == pyads.PLCTYPE_UDINT:
             value = struct.unpack("<I", bytearray(data))[0]
+        elif notification_item.plc_datatype == pyads.PLCTYPE_WORD:
+            value = struct.unpack("<H", bytearray(data))[0]
+        elif notification_item.plc_datatype == pyads.PLCTYPE_DWORD:
+            value = struct.unpack("<I", bytearray(data))[0]
+        elif notification_item.plc_datatype == pyads.PLCTYPE_LREAL:
+            value = struct.unpack("<d", bytearray(data))[0]
+        elif notification_item.plc_datatype == pyads.PLCTYPE_REAL:
+            value = struct.unpack("<f", bytearray(data))[0]
```

Comment:
Can't we simplify this by looking up the way to unpack based on the data type?

Reply:
Thank you, implemented the unpacking based on data type. Strings are handled in a special way.